chain/store/name.go

```diff
@@ -1,63 +1,178 @@
+	return hex.EncodeToString(registrant)
+}
+
+func getNameId(name string, registrant []byte, expiresAt uint32) ([]byte, error) {
+	nameId := bytes.NewBuffer(nil)
+	nameId.Write([]byte(name))
+
+	ni := newNameInfo(registrant, expiresAt)
+	niBytes, err := ni.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	registrantHash := sha256.Sum256(niBytes)
+	nameId.Write(registrantHash[:hashPrefixLength])
+	return nameId.Bytes(), nil
+}
+
+func (sdb *StateDB) getRegName(name string) (string, error) {
+	iter := trie.NewIterator(sdb.trie.NodeIterator(append(NamePrefix, name...)))
+	iter.Next()
+	data := iter.Value
+	if len(data) > 0 {
+		buff := bytes.NewBuffer(data)
+		ni := new(nameInfo)
+		if err := ni.Deserialize(buff); err != nil {
+			return "", err
+		}
+		return ni.String(), nil
+	}
 
-	sdb.names.Store(registrantId, name)
-	sdb.nameRegistrants.Store(nameId, registrant)
+	return "", errors.New("no name registration info found")
 }
 
-func (sdb *StateDB) deleteName(registrantId string) error {
-	name, err := sdb.trie.TryGet(append(NameRegistrantPrefix, registrantId...))
+func (sdb *StateDB) getNameInfo(nameId []byte) (*nameInfo, error) {
+	if v, ok := sdb.names.Load(string(nameId)); ok {
+		if ni, ok := v.(*nameInfo); ok {
+			return ni, nil
+		}
+	}
+
+	ni := new(nameInfo)
+	enc, err := sdb.trie.TryGet(append(NamePrefix, nameId...))
+	if err != nil {
+		return nil, err
+	}
+	if len(enc) > 0 {
+		ni := new(nameInfo)
+		buff := bytes.NewBuffer(enc)
+		if err := ni.Deserialize(buff); err != nil {
+			return nil, err
+		}
+		sdb.names.Store(string(nameId), ni)
+	}
+
+	return ni, nil
+}
+
+func (sdb *StateDB) nameRegistration(name string, registrant []byte, expiresAt uint32) error {
+	nameId, err := getNameId(name, registrant, expiresAt)
 	if err != nil {
 		return err
 	}
 
-	nameId := getNameId(string(name))
-	err = sdb.trie.TryDelete(append(NameRegistrantPrefix, registrantId...))
+	ni, err := sdb.getNameInfo(nameId)
 	if err != nil {
 		return err
 	}
 
-	err = sdb.trie.TryDelete(append(NamePrefix, nameId...))
+	if ni.Empty() {
+		ni.registrant = registrant
+		ni.expiresAt = expiresAt
+	} else {
+		return errors.New("name got registered")
+	}
+
+	return sdb.cleanupNamesAtHeight(expiresAt, nameId)
+}
+
+func (sdb *StateDB) updateNameInfo(nameId string, ni *nameInfo) error {
+	return sdb.updateName(nameId, ni)
+}
+
+func (sdb *StateDB) updateName(nameId string, ni *nameInfo) error {
+	nibytes, err := ni.Bytes()
 	if err != nil {
 		return err
 	}
+	return sdb.trie.TryUpdate(append(NamePrefix, nameId...), nibytes)
+}
 
-	sdb.names.Delete(registrantId)
-	sdb.nameRegistrants.Delete(nameId)
+func (sdb *StateDB) deleteNameInfo(nameId string) error {
+	return sdb.deleteName(nameId)
+}
+
+func (sdb *StateDB) deleteName(id string) error {
+	err := sdb.trie.TryDelete(append(NamePrefix, id...))
+	if err != nil {
+		return err
+	}
 
+	sdb.names.Delete(id)
 	return nil
 }
 
 func (sdb *StateDB) deleteNameForRegistrant(registrant []byte, name string) {
```

Comment:
We no longer need this method because there will be no more pk -> name map stored in db.

Reply:
will keep this method in the old file